qiskit/circuit/quantumcircuit.py

```diff
@@ -1917,6 +1922,26 @@ def num_clbits(self) -> int:
         """Return number of classical bits."""
         return len(self.clbits)
 
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    @name.setter
+    def name(self, name):
+        """Set the name."""
+        self._name = name
+
+    @property
+    def num_params(self):
```

Comment:
This is going to be very confusing along with `num_parameters`. Is there a way to skip this property if the circuit is an `Operation`?

Reply:
@Cryoris, intuitively an `Operation` is what we can add to a `QuantumCircuit`. I believe that our current consensus is that we do want to allow one quantum circuit to contain other quantum circuits as subcircuits (and in particular to allow adding one quantum circuit to another as a subcircuit). One motivating example is that we may have a quantum circuit with two consecutive subcircuits representing Toffoli gates; in this case, we could remove this pair of consecutive Toffolis (something that seems harder to do when the Toffolis are represented using more basic gates). 

Agreed, this duplication `params` vs  `parameters` and `num_params` vs. `num_parameters` is somewhat confusing. At this point, @ShellyGarion and I are not even sure why `Operation` needs `params` in the first place, they were suggested in the original implementation by @kdk; my guess is that it's some kind of extra flexibility that we might need in the future. 

Now, we _could_ simply remove these additional methods from `QuantumCircuit`, since they exist in the parent `Operation` class. However, the method `parameters` in `QuantumCircuit` seems to do additional work. I am not very familiar with this code, but its seems to sort the parameters and to return an object of type `ParameterView`. This, I would expect that the procedure `params` should do the same as `parameters` (which is the reason for the current duplication).

Would be happy to hear any suggestions.

Reply:
I think we should remove `params` from the generic `Instruction` and instead give subclasses an interface that makes more sense. E.g. parameterized gates (ie. with rotation angles) can derive from a `ParameterizedGate` class that adds a `params` (or better `parameters`) property. 

As @kdk suggested below we can replace the current `params` with `operands` which would be what is needed to reconstruct the object like `obj == class(*obj.operands)`.